File: sample/client/proto/PostForumReq.cs
//proto/Forum.proto
using ILib.ProtoPack;
using IWriter = ILib.ProtoPack.IWriter;
using IReader = ILib.ProtoPack.IReader;
using Provider = ILib.ProtoPack.InstanceProvider;

namespace Forum
{
	public partial class PostForumReq : IMessage 
	{
		public Forum.PostData Data { get; set; }

		/// <summary>
		/// Serialize Message
		/// </summary>
		public void Write(IWriter w, bool skipable = true)
		{
			// Write Map Length
			if (!skipable) {
				w.WriteMapHeader(1);
			} else {
				int mapLen = 0;
				if(this.Data != default(Forum.PostData)) mapLen++;
				w.WriteMapHeader(mapLen);
			}
			
			// Write data
			if(!skipable || this.Data != default(Forum.PostData)) {
				var Data = this.Data;
				w.WriteTag(1);
				if (Data == default(Forum.PostData))
				{
					w.WriteNil();
				}
				else
				{
					Data.Write(w, skipable);
				}
			}
		}

		/// <summary>
		/// Deserialize Message
		/// </summary>
		public void Read(IReader r, bool overridable = false)
		{
			// Read Map Length
			var mapLen = r.ReadMapHeader();
			uint tag = 0;
			int index = 0;

			while ((tag = r.ReadTag(index++, mapLen)) != 0)
			{
				switch(tag) {
				case 1:
					if(r.IsNull())
					{
						Data = r.ReadNil<Forum.PostData>();
						continue;
					}
					if(!overridable || Data == default(Forum.PostData))
					{
						Data = Provider.New<Forum.PostData>();
					}
					Data.Read(r, overridable);
					break;
				default:
					r.Skip();
					break;
				}
			}
		}
	}
}

File: sample/server/proto/forum/postforumreq.js
// sample/proto/Forum.proto
"use strict";
const _packer = require('proto-msgpack');
const _proto = require('./../index.js');
class PostForumReq {
	constructor(init, pos) {
		if(init == null || init == true){
			this.data = new _proto.Forum.PostData();
		} else {
			this.data = null;
		}
		if(Buffer.isBuffer(init)){
			this.read(new _packer.ProtoReader(init, pos));
		}
	}
	pack() {
		const w = _packer.defaultWriter;
		w.clear();
		this.write(w);
		return w.toBuffer();
	}
	unpack(buf, pos) {
		if(!Buffer.isBuffer(buf)){
			this.read(buf);
		} else {
			this.read(new _packer.ProtoReader(buf, pos));
		}
	}
	write(w) {
		// Write Map Length
		w.writeMapHeader(1);
		
		// Write data
		w.writeTag(1);
		if (!this.data) {
			w.writeNil();
		} else {
			this.data.write(w);
		}
	}
	read(r) {
		// Read Map Length
		const mapLen = r.readMapHeader();

		for(let i = 0; i < mapLen; i++) {
			const tag = r.readTag();
			switch(tag) {
			case 1:
				if(r.isNull()) {
					r.readNil();
					this.data = null;
				} else {
					this.data = new _proto.Forum.PostData(false);
					this.data.read(r);
				}
				break;
			default:
				r.skip();
				break;
			}
		}
	}
}
module.exports = PostForumReq
